atc/exec/task_step.go

```diff
@@ -190,16 +195,57 @@ func (step *TaskStep) Run(ctx context.Context, state RunState) error {
 
 	owner := db.NewBuildStepContainerOwner(step.metadata.BuildID, step.planID, step.metadata.TeamID)
 
-	chosenWorker, err := step.workerPool.FindOrChooseWorkerForContainer(
-		ctx,
-		logger,
-		owner,
-		containerSpec,
-		workerSpec,
-		step.strategy,
-	)
-	if err != nil {
-		return err
+	var chosenWorker worker.Worker
+	var activeTasksLock lock.Lock
+
+	for {
+		if step.strategy.ModifiesActiveTasks() {
+			var acquired bool
+			activeTasksLock, acquired, err = step.lockFactory.Acquire(logger, lock.NewTaskStepLockID())
+			if err != nil {
+				return err
+			}
+
+			if !acquired {
+				time.Sleep(time.Second)
+				continue
+			}
+		}
+
+		chosenWorker, err = step.workerPool.FindOrChooseWorkerForContainer(
+			ctx,
+			logger,
+			owner,
+			containerSpec,
+			workerSpec,
+			step.strategy,
+		)
+		if err != nil {
+			return err
+		}
+
+		if step.strategy.ModifiesActiveTasks() {
+			if chosenWorker == nil {
```

Comment:
About the first comment:
That's already been taken care of as is because `pool.allSatisfying` already returns an error when the pool is empty, so that behavior is unchanged.

About the timeout: I've thought about it. If we are to implement it should be customizable and maybe that's the right approach eventually, but for now I really wanted to keep the PR as simple as possible. We can refine the feature incrementally afterward.

Reply:
Thx @aledeganopix4d